---
title: "R.Homework"
author: "Sobhan Bahrami"
date: "5/3/2022"
output: github_document
---

```{r setup, include=FALSE}
#install.packages("ggpmisc")  
library(tidyverse)
library(ggplot2)
library(ggpmisc)
knitr::opts_chunk$set(echo = TRUE)

```

##  Homework for Module 1.1: R

For my homework, I used the results of one of my recent projects. In this project, I tried to analyze RNA-SEQ data for a specific species. Our aim was to compare the expression level of different genes between different treatments. So, the file that I used for this homework was generated by the HT-Seq tool. In this file, we have information about the numbers of reads for each gene for four samples (two treatments and two replications per treatment). There are around 15500 genes for this specific species and we expect to see consistent data between replications for each treatment. My aim is to calculate the slope of the linear model for each treatment to see if they are consistent or not. The data that I am using is not normalized and I have to first normalize it according to the total read count for each sample. Also, the lengths of genes are important because we will have more reads for larger genes. In the input file, we have a column that shows if the gene is large (L) (>3500bp) or small (S)(<=3500bp).

The first step is to import the read count file in R. For this step, I used the following command:
```{r read counts}
read_counts <- read_csv("/home/sobhan/Desktop/Bowtie_aligned_>=1_time_without_normalization.txt")
print(read_counts)
```

Now, I want to calculate the average of read counts in each sample. Therefore, I used the following command for this purpose:

```{r mean read counts}
Average<-c(Two_Four_c_rep1=mean(read_counts$Two_Four_c_rep1),Two_Four_c_rep2=mean(read_counts$Two_Four_c_rep2),Four_Eight_c_rep1=mean(read_counts$Four_Eight_c_rep1),Four_Eight_c_rep2=mean(read_counts$Four_Eight_c_rep2))
Average

```
According to this result, we first need to normalize our data. However, I also want to calculate the average of read counts for large and small genes in each sample. To do so, I used the following command:

```{r mean read counts for large and small genes}
read_counts_summ <- read_counts %>%
  group_by(Length) %>%
  summarise(mean.Two_Four_c_rep1 = mean(Two_Four_c_rep1, na.rm = TRUE),
            mean.Two_Four_c_rep2 = mean(Two_Four_c_rep2, na.rm = TRUE),
            mean.Four_Eight_c_rep1 = mean(Four_Eight_c_rep1, na.rm = TRUE),
            mean.Four_Eight_c_rep2 = mean(Four_Eight_c_rep2, na.rm = TRUE))
read_counts_summ 

```
According to these results, as we expected, the average of the read counts for larger genes is more than the mean read counts for small genes in each sample. Now, I want to normalize the data according to the total read count for each sample. So here I calculated the total read count for each sample and then calculated the normalized read counts for each sample. So, after this step, the total read count for each sample will be 1.000.000.

```{r normalization}
read_counts_norm <- read_counts %>%
  group_by(Length) %>%
  mutate(norm.Two_Four_c_rep1 = Two_Four_c_rep1 * (1000000/sum(Two_Four_c_rep1, na.rm = TRUE)),
         norm.Two_Four_c_rep2 = Two_Four_c_rep2 * (1000000/sum(Two_Four_c_rep2, na.rm = TRUE)),
         norm.Four_Eight_c_rep1 = Four_Eight_c_rep1 * (1000000/sum(Four_Eight_c_rep1, na.rm = TRUE)),
         norm.Four_Eight_c_rep2 = Four_Eight_c_rep2 * (1000000/sum(Four_Eight_c_rep2, na.rm = TRUE)))
glimpse(read_counts_norm)
read_counts_norm
```
The next step is to create a list column of our data frame. So, I did this step by using `nest()` and then I defined a function to calculate a linear regression for each treatment. In this step `lm()` function was used. Also, to get the slope of the linear model for each treatment I wrote a new function, `get_slope`.

These steps should be repeated for both of our treatments. First, I did it for the "Four_Eight_c" treatment:

```{r linear model}
nested_df <- read_counts_norm %>%
  group_by(Length) %>%
  nest()
#nested_df$data[[1]]
glimpse(nested_df$data[[1]])
glimpse(nested_df)
lm_fun <- function(df) {
  lm(norm.Four_Eight_c_rep1 ~ norm.Four_Eight_c_rep2, data = df)
}
nested_df_2.2 <- nested_df %>%
  mutate(lms = map(data, lm_fun))
summary(nested_df_2.2$lms[[1]])
get_slope <- function(my.lm) {
  coef(my.lm)[2]
}

slopes <- nested_df_2.2 %>%
  mutate(slope = map_dbl(lms, get_slope))
print(slopes)
```
According to these results, we can see that the slopes of the linear models in the Four_Eight_c treatment for both large and small genes are close to each other and for small genes is near to 1, which is perfect. 

And again I repeated the previous steps for the "Two_Four_c" treatment:

```{r linear model2}
nested_df <- read_counts_norm %>%
  group_by(Length) %>%
  nest()
#nested_df$data[[1]]
glimpse(nested_df$data[[1]])
glimpse(nested_df)
lm_fun <- function(df) {
  lm(norm.Two_Four_c_rep1 ~ norm.Two_Four_c_rep2, data = df)
}
nested_df_2.1 <- nested_df %>%
  mutate(lms = map(data, lm_fun))
summary(nested_df_2.1$lms[[1]])
get_slope <- function(my.lm) {
  coef(my.lm)[2]
}

slopes <- nested_df_2.1 %>%
  mutate(slope = map_dbl(lms, get_slope))
print(slopes)
```
And also we can see that for this treatment the slopes of the linear models are close to 1 which is good.

## Plots

Now I want to use ggplot2 to show my data and compare the replications together more easily

For "Two_Four_c" treatment:


```{r Plot 1}
#png(file="/home/sobhan/Desktop/2_4_c_rep1.VS.2_4_c_rep2.png", width=1200, height=900)
ggplot(nested_df$data[[2]], aes(x=norm.Two_Four_c_rep1, y=norm.Two_Four_c_rep2)) +
  geom_point(shape=18, color="orange")+
  #geom_text(label=read_counts$Gene)+
  geom_smooth(method=lm, se=FALSE, linetype="dashed", color="blue")+ geom_rug()+labs(x = expression(italic(Two_Four_c_rep1)), y = expression(italic(Two_Four_c_rep2)))
```


And for the "Four_Eight_c" treatment:

```{r Plot 2}

#png(file="/home/sobhan/Desktop/4_8_c_rep1.VS.4_8_c_rep2.png", width=1200, height=900)
ggplot(nested_df$data[[1]], aes(x=norm.Four_Eight_c_rep1, y=norm.Four_Eight_c_rep2)) +
  geom_point(shape=18, color="orange")+
  #geom_text(label=read_counts$Gene)
  geom_smooth(method=lm, se=FALSE, linetype="dashed", color="blue")+ geom_rug()+
  labs(x = expression(italic(Four_Eight_c_rep1)), y = expression(italic(Four_Eight_c_rep2)))

```

So these results indicate that the normalization step was done correctly and our replications for each treatment are consistent. For our study, in which we want to compare the read counts for genes between different treatments, this kind of normalization is suitable. However, we can also consider the lengths of the genes and try to normalize the data according to their lengths. This step may help us to make our data more consistent.
